docs/writing-controllers.md

```diff
@@ -0,0 +1,1489 @@
+  never,
+  FooControllerEvents,
+  never,
+  FooControllerEvents['type']
+>;
+```
+
+✅ **`never` is passed as the 5th type parameter (assuming no external events)**
+
+```typescript
+export type FooControllerEvents =
+  | FooControllerMessageReceivedEvent
+  | FooControllerNotificationAddedEvent;
+
+export type FooControllerMessenger = RestrictedControllerMessenger<
+  'FooController',
+  never,
+  FooControllerEvents,
+  never,
+  never
+>;
+```
+
+## Define external action types as one, but do not export them
+
+A controller may wish to call actions defined by other controllers, and therefore will need to define them in the messenger's allowlist.
+
+In this case, the controller should group these types into a type union so that they can be easily passed to the `RestrictedControllerMessenger` type. However, it should not export this type, as it would then be re-exporting types that another package has already exported.
+
+The name of this type should be `AllowedActions`.
+
+This type should not only be passed to `RestrictedControllerMessenger` as the 2nd type parameter, but should also be included in its 4th type parameter.
+
+🚫 **`never` is passed as the 4th type parameter**
+
+```typescript
+export type AllowedActions =
+  | BarControllerDoSomethingAction
+  | BarControllerDoSomethingElseAction;
+
+export type FooControllerMessenger = RestrictedControllerMessenger<
+  'FooController',
+  AllowedActions,
+  never,
+  never,
+  never
+>;
+```
+
+🚫 **`AllowedActions['type']` is passed as the 4th type parameter, but `AllowedActions` is exported**
+
+```typescript
+/* === packages/foo-controller/src/FooController.ts === */
+
+export type AllowedActions =
+  | BarControllerDoSomethingAction
+  | BarControllerDoSomethingElseAction;
+
+export type FooControllerMessenger = RestrictedControllerMessenger<
+  'FooController',
+  AllowedActions,
+  never,
+  AllowedActions['type'],
+  never
+>;
+
+/* === packages/foo-controller/src/index.ts === */
+
+export type { AllowedActions } from '@metamask/foo-controller';
+```
+
+✅ **`AllowedActions['type']` is passed as the 4th type parameter, and `AllowedActions` is _not_ exported**
+
+```typescript
+type AllowedActions =
+  | BarControllerDoSomethingAction
+  | BarControllerDoSomethingElseAction;
+
+export type FooControllerMessenger = RestrictedControllerMessenger<
+  'FooController',
+  AllowedActions,
+  never,
+  AllowedActions['type'],
+  never
+>;
+```
+
+If, in a test, you need to access all of the actions included in a controller's messenger allowlist, use the [`ExtractAvailableAction` utility type](../packages/base-controller/tests/helpers.ts):
+
+```typescript
+// NOTE: You may need to adjust the path depending on where you are
+import { ExtractAvailableAction } from '../../base-controller/tests/helpers';
+
+const messenger = new ControllerMessenger<
+  ExtractAvailableAction<FooControllerMessenger>,
+  never
+>();
+```
+
+## Define external event types as one, but do not export them
```

Comment:
For consistency with https://github.com/MetaMask/core/pull/4110#discussion_r1670856411

```suggestion
## Define, but do not export, a type union for external event types
```